### PR TITLE
[Feature] CSV import and export support header

### PR DESCRIPTION
**Pull request by chenlinzhong**
# Proposed changes

Issue Number: open #8487 

## Problem Summary:

- Add two new types to stream load boker load: **csv_with_names** and **csv_with_name_sand_types**
- Add two new types to export: **csv_with_names** and **csv_with_names_and_types**

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/http/action/stream_load.cpp [hunk 1]
```diff
@@ -265,8 +266,14 @@ Status StreamLoadAction::_on_header(HttpRequest* http_req, StreamLoadContext* ct
...
         return Status::InternalError("compress data of JSON format is not supported.");
     }
-    ctx->format =
-            parse_format(http_req->header(HTTP_FORMAT_KEY), http_req->header(HTTP_COMPRESS_TYPE));
+    std::string format_str = http_req->header(HTTP_FORMAT_KEY);
+    if (boost::iequals(format_str, BeConsts::CSV_WITH_NAMES) ||
+        boost::iequals(format_str, BeConsts::CSV_WITH_NAMES_AND_TYPES)) {
+        ctx->header_type = format_str;
+        //treat as CSV
+        format_str = "csv";
```
Comment (cambyzju) [inline be/src/http/action/stream_load.cpp:274]: "csv" also use an const define like CSV_WITH_NAMES?

File: fe/fe-core/src/main/java/org/apache/doris/load/BrokerFileGroup.java [hunk 2]
```diff
@@ -214,7 +215,9 @@ public void parse(Database db, DataDescription dataDescription) throws DdlExcept
             if (!fileFormat.equalsIgnoreCase("parquet")
                     && !fileFormat.equalsIgnoreCase("csv")
```
Comment (cambyzju) [inline fe/fe-core/src/main/java/org/apache/doris/load/BrokerFileGroup.java:216]: "csv" define as const like csv_with_names

File: be/src/http/action/stream_load.cpp (hunk 1 above)
Comment (chenlinzhong) [reply be/src/http/action/stream_load.cpp:274]: ok

Comment (chenlinzhong) [reply be/src/http/action/stream_load.cpp:274]: ok

File: be/src/runtime/export_sink.cpp [hunk 3]
```diff
@@ -96,6 +96,13 @@ Status ExportSink::send(RuntimeState* state, RowBatch* batch) {
     int num_rows = batch->num_rows();
     // we send at most 1024 rows at a time
     int batch_send_rows = num_rows > 1024 ? 1024 : num_rows;
+    if (_t_export_sink.header.size() > 0) {
+        size_t written_len = 0;
+        RETURN_IF_ERROR(
+                _file_writer->write(reinterpret_cast<const uint8_t*>(_t_export_sink.header.c_str()),
+                                    _t_export_sink.header.size(), &written_len));
+        _t_export_sink.header = "";
```
Comment (yinzhijian) [inline be/src/runtime/export_sink.cpp:104]: use a flag  to indicate that the header has been sent may be better?

Comment (chenlinzhong) [reply be/src/runtime/export_sink.cpp:104]: ok

File: fe/fe-core/src/main/java/org/apache/doris/load/BrokerFileGroup.java (hunk 2 above)
Comment (chenlinzhong) [reply fe/fe-core/src/main/java/org/apache/doris/load/BrokerFileGroup.java:216]: ok

File: be/src/runtime/export_sink.cpp [hunk 4]
```diff
@@ -96,6 +97,13 @@ Status ExportSink::send(RuntimeState* state, RowBatch* batch) {
     int num_rows = batch->num_rows();
     // we send at most 1024 rows at a time
     int batch_send_rows = num_rows > 1024 ? 1024 : num_rows;
+    if (!_send_header && _t_export_sink.header.size() > 0) {
+        size_t written_len = 0;
+        RETURN_IF_ERROR(
+                _file_writer->write(reinterpret_cast<const uint8_t*>(_t_export_sink.header.c_str()),
+                                    _t_export_sink.header.size(), &written_len));
+        _send_header = true;
```
Comment (yinzhijian) [inline be/src/runtime/export_sink.cpp:105]: use _header_sent may be better

File: be/src/runtime/export_sink.h [hunk 5]
```diff
@@ -77,6 +77,7 @@ class ExportSink : public DataSink {
     RuntimeProfile::Counter* _bytes_written_counter;
     RuntimeProfile::Counter* _rows_written_counter;
     RuntimeProfile::Counter* _write_timer;
+    bool _send_header;
```
Comment (yinzhijian) [inline be/src/runtime/export_sink.h:80]: _header_sent may be better

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md [hunk 6]
```diff
@@ -78,7 +78,7 @@ under the License.
 
         exec_mem_limit: 导入内存限制。默认为 2GB。单位为字节。
 
-        format: 指定导入数据格式，默认是csv，支持json格式。
+        format: 指定导入数据格式，默认是csv，支持json格和带header的csv格式: csv_with_names、csv_with_names_and_types。
```
Comment (yinzhijian) [inline docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md:81]: 支持json格式， 少个式

File: fe/fe-core/src/main/java/org/apache/doris/load/ExportJob.java [hunk 7]
```diff
@@ -252,10 +253,40 @@ private void genExecFragment() throws UserException {
...
+                header = header.substring(0, header.length() - getColumnSeparator().length());
+                header = header + getLineDelimiter();
+            } else if (headerType.equals(FeConstants.csv_with_names_and_types)) {
+                //names
+                for (SlotDescriptor slot : exportTupleDesc.getSlots()) {
+                    header = header + slot.getColumn().getName() + getColumnSeparator();
+                }
+                header = header.substring(0, header.length() - getColumnSeparator().length());
+                header = header + getLineDelimiter();
+                //types
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/load/ExportJob.java:279]: use a common function to gen names may be better

File: fe/fe-core/src/main/java/org/apache/doris/planner/ExportSink.java [hunk 8]
```diff
@@ -44,6 +45,15 @@ public ExportSink(String exportPath, String columnSeparator,
         this.brokerDesc = brokerDesc;
     }
 
+    public ExportSink(String exportPath, String columnSeparator,
+                      String lineDelimiter, BrokerDesc brokerDesc, String header) {
+        this.exportPath = exportPath;
+        this.columnSeparator = columnSeparator;
+        this.lineDelimiter = lineDelimiter;
+        this.brokerDesc = brokerDesc;
```
Comment (yinzhijian) [inline fe/fe-core/src/main/java/org/apache/doris/planner/ExportSink.java:53]: ExportSink(exportPath, columnSeparator, lineDelimiter, brokerDesc); may be better

Comment (chenlinzhong) [reply fe/fe-core/src/main/java/org/apache/doris/planner/ExportSink.java]: ok

**Review by yinzhijian — COMMENTED** [review]
LGTM

File: be/src/runtime/file_result_writer.cpp [hunk 9]
```diff
@@ -216,12 +218,82 @@ std::string FileResultWriter::_file_format_to_name() {
...
     }
 }
+std::string FileResultWriter::gen_types() {
+    std::string types = "";
+    int num_columns = _output_expr_ctxs.size();
+    for (int i = 0; i < num_columns; ++i) {
+        switch (_output_expr_ctxs[i]->root()->type().type) {
+        case TYPE_BOOLEAN:
+        case TYPE_TINYINT:
+            types += "TINYINT";
```
Comment (cambyzju) [inline be/src/runtime/file_result_writer.cpp:228]: 1. can we use type_to_string in primitive_type.cpp?
2. the string for types, it is better the same as types name in create table SQL?

Comment (chenlinzhong) [reply be/src/runtime/file_result_writer.cpp]: ok, it's a good idea

**Comment by cambyzju** [comment]
+1

File: be/src/runtime/export_sink.cpp (hunk 4 above)
Comment (chenlinzhong) [reply be/src/runtime/export_sink.cpp]: ok

File: be/src/runtime/export_sink.h (hunk 5 above)
Comment (chenlinzhong) [reply be/src/runtime/export_sink.h]: ok

File: docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md (hunk 6 above)
Comment (chenlinzhong) [reply docs/zh-CN/sql-reference/sql-statements/Data Manipulation/STREAM LOAD.md]: ok

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**